README.rst
========
subparse
========

A wrapper for argparse that provides decorator-based subcommand support.

Commands can be defined separately from their actual main functions,
enabling faster import times.

Basic Usage
===========

::

    from subparse import CLI

    class MyApp(object):
        def __init__(self, quiet=False):
            self.quiet = quiet

        def info(self, *msg):
            if not self.quiet:
                print('[info]', *msg)

    def context_factory(cli, args):
        return MyApp(args.quiet)

    def generic_options(parser):
        parser.add_argument('--quiet', action='store_true',
                            help='turn of debugging')

    cli = CLI(version='0.0', context_factory=context_factory)
    cli.add_generic_options(generic_options)

    @cli.command(__name__ + ':foo_main')
    def foo(parser):
        """
        a short description ending in a period.

        a longer description
        """
        parser.add_argument('--bar', action='store_true',
                            help='turn on bar')

    def foo_main(app, args):
        app.info('Hello World!')

    result = cli.run()
    sys.exit(result)

Lazy Decorators
===============

Commands can be defined lazily and picked up later. This removes ordering
restrictions between the commands and the cli object.

A module containing commands can be defined irrespective of the actual
``CLI`` instance:

::

    # myapp/info.py

    from subparse import command

    @command('myapp.info:foo_main')
    def foo(parser):
        """perform foo"""

Later, when an instance of a ``CLI`` is created, the commands can be loaded
and registered:

::

    cli = CLI()
    cli.load_commands('myapp.info')

Entry Points
============

Commands may also be defined in external modules and loaded via entry
points.

::

    from subparse import cli

    cli = CLI()
    cli.load_commands_from_entry_point('myapp.commands')

An extension application would then define the external module that should
be searched for commands. Again this allows the commands themselves to be
defined independently of the main functions, improving import speed.

An extension package should define a module containing the supported
commands:

::

    # barpkg/commands.py

    from subparse import command

    @command('barpkg.bar')
    def bar(parser):
        """perform bar"""

The package should also define the function to be called for each command.
Optionally in a separate module to avoid importing run-time dependencies
during parsing:

::

    # barpkg/bar.py

    def main(app, args):
        pass

The package can then broadcast the module ``barpkg.commands``
containing the supported commands:

::

    [myapp.commands]
    barpkg = barpkg.commands

Now when your extension package is installed the commands will automatically
become available.

Context Factory
===============

Each subcommand, when executed, is passed a context object which defines a
reusable API between subcommands. This is really the secret sauce of
``subparse`` that makes it really easy to build your own shared CLI features.

The ``context_factory`` argument to the ``subparse.CLI`` allows for defining
an object that is passed to all commands. This factory can also be a
generator, allowing it to ``yield`` the context object and then cleanup
after the command is complete. For example:

::

    import transaction

    def context_factory(cli, args):
        tm = transaction.TransactionManager(explicit=True)
        with tm:
            yield tm

In the above example the transaction manager is available to all subcommands
and it can commit/abort based on whether the command raises an exception.

Each subcommand can pass custom kwargs to the context factory via the
``context_kwargs`` argument. For example, if a single subcommand wishes to
opt-out of the transaction manager:

::

    def context_factory(cli, args, without_tm=False):
        if without_tm:
            yield

        tm = transaction.TransactionManager(explicit=True)
        with tm:
            yield tm

    @command(..., context_kwargs=dict(without_tm=True))
    def foo(parser):
        """" Run a command without the tm enabled."""


CHANGES.rst
0.6 (2022-05-15)
================

- Drop Python 2.7, 3.4, 3.5, 3.6, 3.7.

- Add Python 3.8, 3.9, 3.10.

- Drop dependency on pkg_resources and use importlib.metadata.

- 100% test coverage.

0.5.3 (2019-03-09)
==================

- Output help to ``sys.stderr`` when parsing fails.

- Support passing ``context_kwargs`` to the ``command`` decorator. These
  arguments will be passed to the ``context_factory`` when the command is
  executed.

0.5.2 (2019-03-09)
==================

- Sort subcommands in help output.

0.5.1 (2019-03-08)
==================

- Use the ``argparse.RawTextHelpFormatter`` formatter class.

0.5 (2019-03-08)
================

- Add Python 3.7 support.

- Fix a deprecation warning coming from setuptools.

- Conform more closely to PEP-257 for docstring parsing.

- Modify how the help text is displayed using the
  ``argparse.RawDescriptionHelpFormatter`` formatter class.

0.4 (2018-05-03)
================

- Drop Python 2.6, 3.2 and 3.3 support.

- Add Python 3.4, 3.5, 3.6 support.

- Allow the ``context_factory`` to be a generator which yields the context.
  This allows the context to wrap the full lifecycle of the CLI.

0.3.3 (2013-08-12)
==================

No functional changes from 0.3.2.

- Improve documentation.

0.3.2 (2013-08-06)
==================

- Add `CLI.run` API for simply executing the command line.

0.3.1 (2013-08-06)
==================

- Improve the help output.

0.3 (2013-08-06)
================

- Rename subcommands to commands in the API.

0.2 (2013-08-06)
================

- Underscores in function names are converted to dashes in their respective
  subcommand names.
- Add `CLI.add_generic_options` API.
- Add a new `help` subcommand, allowing for `myapp help foo`.
- Allow relative imports in the subcommand specification.

0.1 (2013-08-05)
================

- Initial Commits


tests/__init__.py
import os
import sys

here = os.path.dirname(__file__)
packages_dir = os.path.join(here, 'fake_packages')
for pkg in os.listdir(packages_dir):
    sys.path.insert(0, os.path.join(packages_dir, pkg))


tests/test_cli.py
import pytest


def make_cli(*args, **kwargs):
    from subparse import CLI

    context_factory = kwargs.pop('context_factory', None)
    if context_factory is None:
        context = kwargs.pop('context', None)
        context_factory = lambda a, b: context

    return CLI(*args, context_factory=context_factory, **kwargs)


def test_basic():
    app = {}
    cli = make_cli(context=app)

    @cli.command('.fixtures.foo')
    def foo(parser):
        """
        Short name

        This is my long description.
        """
        parser.add_argument('--bar', action='store_true')

    result = cli.run(['foo', '--bar'])
    assert app['bar'] is True
    assert result == 0


def test_context_generator():
    out = []

    def context_factory(cli, args):
        out.append(1)
        yield {}
        out.append(2)

    cli = make_cli(context_factory=context_factory)

    @cli.command('.fixtures.foo')
    def foo(parser):
        parser.add_argument('--bar', action='store_true')

    result = cli.run(['foo', '--bar'])
    assert result == 0
    assert out == [1, 2]


def test_basic_decorator_dict():
    from subparse import command

    @command('.fixtures.foo')
    def foo(parser):
        parser.add_argument('--bar', action='store_true')

    app = {}
    cli = make_cli(context=app)
    cli.load_commands(locals())

    result = cli.run(['foo', '--bar'])
    assert app['bar'] is True
    assert result == 0


def test_load_module():
    from .fixtures import foo

    app = {}
    cli = make_cli(context=app)
    cli.load_commands(foo)

    result = cli.run(['bar'])
    assert app['fn'] == 'bar_main'
    assert app['bar'] is False
    assert result == 0


def test_load_module_dotted():
    cli = make_cli()
    cli.load_commands('.fixtures.foo')
    assert len(cli.commands) > 0


def test_load_module_dotted_relative():
    cli = make_cli()
    cli.load_commands('.fixtures.foo')
    assert len(cli.commands) > 0


def test_main_callable():
    from .fixtures.foo import main

    app = {}
    cli = make_cli(context=app)

    @cli.command(main)
    def foo(parser):
        parser.add_argument('--bar', action='store_true')

    result = cli.run(['foo', '--bar'])
    assert app['fn'] == 'main'
    assert app['bar'] is True
    assert result == 0


def test_main_dotted_relative():
    app = {}
    cli = make_cli(context=app)

    @cli.command('.fixtures.foo:foo_main')
    def foo(parser):
        parser.add_argument('--bar', action='store_true')

    result = cli.run(['foo', '--bar'])
    assert app['fn'] == 'foo_main'
    assert app['bar'] is True
    assert result == 0


def test_underscore_command_name_converted_to_dash():
    app = {}
    cli = make_cli(context=app)

    @cli.command('.fixtures.foo')
    def foo_bar(parser):
        parser.add_argument('--bar', action='store_true')

    result = cli.run(['foo-bar', '--bar'])
    assert app['bar'] is True
    assert result == 0


def test_underscore_command_name_preserved():
    app = {}
    cli = make_cli(context=app)

    @cli.command('.fixtures.foo', 'foo_bar')
    def foo_bar(parser):
        parser.add_argument('--bar', action='store_true')

    result = cli.run(['foo_bar', '--bar'])
    assert app['bar'] is True
    assert result == 0


def test_override_command_name():
    app = {}
    cli = make_cli(context=app)

    @cli.command('.fixtures.foo', 'baz')
    def foo_bar(parser):
        parser.add_argument('--bar', action='store_true')

    result = cli.run(['baz', '--bar'])
    assert app['bar'] is True
    assert result == 0


def test_alternate_main():
    app = {}
    cli = make_cli(context=app)

    @cli.command('.fixtures.foo:foo_main')
    def foo(parser):
        parser.add_argument('--bar', action='store_true')

    result = cli.run(['foo', '--bar'])
    assert app['fn'] == 'foo_main'
    assert app['bar'] is True
    assert result == 0


def test_blank_command(capsys):
    cli = make_cli()

    @cli.command('.fixtures.foo')
    def foo(parser):
        parser.add_argument('--bar', action='store_true')

    pytest.raises(SystemExit, cli.run, [])
    out, err = capsys.readouterr()
    assert 'usage:' in err


def test_help_command(capsys):
    cli = make_cli()

    @cli.command('.fixtures.foo')
    def foo(parser):
        parser.add_argument('--bar', action='store_true')

    pytest.raises(SystemExit, cli.run, ['help', 'foo'])
    out, err = capsys.readouterr()
    assert 'usage:' in out


def test_bad_options(capsys):
    cli = make_cli()

    @cli.command('.fixtures.foo')
    def foo(parser):
        parser.add_argument('--bar', action='store_true')

    pytest.raises(SystemExit, cli.run, ['foo', '--missing'])
    out, err = capsys.readouterr()
    assert 'usage:' in err


def test_version_default(capsys):
    cli = make_cli(version='foo')

    pytest.raises(SystemExit, cli.run, ['--version'])
    out, err = capsys.readouterr()
    assert out + err == 'foo\n'


def test_docstring(capsys):
    from subparse import parse_docstring

    short, long = parse_docstring(
        '''
        short1
        short2

        long1
        long2
        '''
    )
    assert short == 'short1 short2'
    assert long == 'long1\nlong2'

    short, long = parse_docstring(
        '''hello world
        this is part of short

        this is part of long
        '''
    )
    assert short == 'hello world this is part of short'
    assert long == 'this is part of long'


def test_load_entry_point():
    app = {}
    cli = make_cli(context=app)
    cli.load_commands_from_entry_point('cli.commands')

    result = cli.run(['foo'])
    assert app['fn'].__module__ == 'fakeapp.foo'
    assert app['fn'].__name__ == 'main'
    assert app['args'].bar is False
    assert result == 0


tests/fixtures/foo.py
from subparse import command


@command(__name__)
def foo(parser):
    parser.set_defaults(call='foo')
    parser.add_argument('--bar', action='store_true')


@command('.')
def foo_main_dot(parser):
    parser.set_defaults(call='foo_main_dot')
    parser.add_argument('--bar', action='store_true')


@command(__name__ + ':foo_main')
def foo_main_absolute_colon(parser):
    parser.set_defaults(call='foo_main_absolute_colon')
    parser.add_argument('--bar', action='store_true')


@command(__name__ + '.foo_main')
def foo_main_dotted(parser):
    parser.set_defaults(call='foo_main_dotted')
    parser.add_argument('--bar', action='store_true')


@command('.foo_main')
def foo_main_relative_leading_dot(parser):
    parser.set_defaults(call='foo_main_relative_leading_dot')
    parser.add_argument('--bar', action='store_true')


@command(':foo_main')
def foo_main_relative_leading_colon(parser):
    parser.set_defaults(call='foo_main_relative_leading_colon')
    parser.add_argument('--bar', action='store_true')


@command('.foo:bar_main', 'bar')
def bar_options(parser):
    parser.set_defaults(call='bar_options')
    parser.add_argument('--bar', action='store_true')


def main(app, args):
    app['fn'] = 'main'
    app['call'] = getattr(args, 'call', None)
    app['bar'] = args.bar


def foo_main(app, args):
    app['fn'] = 'foo_main'
    app['call'] = getattr(args, 'call', None)
    app['bar'] = args.bar


def bar_main(app, args):
    app['fn'] = 'bar_main'
    app['call'] = getattr(args, 'call', None)
    app['bar'] = args.bar


src/subparse/lazydecorator.py
"""
lazy decorators:  remove global state from your app, e.g. Bottle and Flask.

A lazy decorator takes the place of another decorator, but just memoizes
decoration parameters and lets you choose when to apply the actual decorator.
This means that you are not tied to apply decorators like the typical
flask/bottle ``@app.route("/index")`` at import time and thus don't
need to create a global ``app`` object.

Example usage in a module.py:

    from lazydecorator import lazydecorator

    route = lazydecorator()

    class MyServer:
        @route("/index")
        def index(self):
            pass

The lazydecorator "route" instance returns the same ``index`` function it
receives but sets an attribute to remember the ``("/index")`` parameter.
Later, after importing the ``module`` you can then apply your  ``@app.route``
decorator like this::

    def create_app():
        app = Bottle()
        import module
        myserver = module.MyServer()
        module.route.discover_and_call(myserver, app.route)
        # The app.route decorator is called with the bound
        # ``myserver.index`` method

order of registrations is preserved.

(c) holger krekel, 2013, License: MIT
"""
from types import FunctionType


class lazydecorator:
    def __init__(self):
        self.attrname = "_" + hex(id(self))
        self.num = 0

    def __call__(self, *args, **kwargs):
        def decorate(func):
            try:
                num, siglist = getattr(func, self.attrname)
            except AttributeError:
                siglist = []
                func.__dict__[self.attrname] = (self.num, siglist)
                self.num += 1
            siglist.append((args, kwargs))
            return func

        return decorate

    def discover(self, obj):
        decitems = []
        if isinstance(obj, dict):

            def iter():
                for name in obj:
                    yield name, obj[name]

        else:

            def iter():
                for name in dir(obj):
                    yield name, getattr(obj, name)

        for _, func in iter():
            func_orig = func
            if not isinstance(func, FunctionType):
                try:
                    func = func.__func__
                except AttributeError:
                    continue
            try:
                num, siglist = getattr(func, self.attrname)
            except AttributeError:
                continue
            decitems.append((num, func_orig, siglist))
        decitems.sort()
        result = []
        for _, func_orig, siglist in decitems:
            for args, kwargs in siglist:
                result.append((func_orig, args, kwargs))
        return result

    def discover_and_call(self, obj, dec):
        for func, args, kwargs in self.discover(obj):
            newfunc = dec(*args, **kwargs)(func)
            assert newfunc == func


src/subparse/__init__.py
import argparse
from collections import namedtuple
from contextlib import contextmanager
from importlib import import_module
import importlib.metadata
import inspect
import sys

from .lazydecorator import lazydecorator

command = lazydecorator()

CommandMeta = namedtuple(
    'CommandMeta',
    ['factory', 'main', 'name', 'help', 'description', 'context_kwargs'],
)


class ArgumentParser(argparse.ArgumentParser):
    def error(self, message):
        """Raise errors instead of printing and raising SystemExit."""
        raise argparse.ArgumentError(None, message)


class CLI:
    _ArgumentParser = ArgumentParser
    _namespace_key = '_subparse_meta'

    def __init__(
        self,
        prog=None,
        usage=None,
        description=None,
        version=None,
        add_help_command=True,
        context_factory=None,
    ):
        self.prog = prog
        self.usage = usage
        self.description = description
        self.version = version
        self.add_help_command = add_help_command
        self.generic_options = []
        self.commands = {}
        self.context_factory = context_factory

        if version is not None:
            self.add_generic_option(
                '-V', '--version', action='version', version=version
            )

    def add_generic_options(self, generic_options):
        """
        Register a function containing generic options.

        The function should accept an instance of an
        :class:`argparse.ArgumentParser` and use it to define extra
        arguments and options.

        """
        self.generic_options.append(generic_options)

    def add_generic_option(self, *args, **kwargs):
        def generic_options(parser):
            parser.add_argument(*args, **kwargs)

        self.add_generic_options(generic_options)

    def add_command(self, factory, main, name=None, context_kwargs=None):
        """
        Attach a command directly to the :class:`CLI` object.

        """
        if name is None:
            name = factory.__name__.replace('_', '-')

        if context_kwargs is None:
            context_kwargs = {}

        short_desc, long_desc = parse_docstring(factory.__doc__)
        if long_desc:
            long_desc = short_desc + '\n\n' + long_desc

        # determine the absolute import string if relative
        if isinstance(main, str) and (main.startswith('.') or main.startswith(':')):
            module = __import__(factory.__module__, None, None, ['__doc__'])
            package = package_for_module(module)
            if main in ['.', ':']:
                main = package.__name__
            else:
                main = package.__name__ + main

        self.commands[name] = CommandMeta(
            factory=factory,
            main=main,
            name=name,
            help=short_desc,
            description=long_desc,
            context_kwargs=context_kwargs,
        )

    def command(self, *args, **kwargs):
        """
        Attach a command to the current :class:`CLI` object.

        The function should accept an instance of an
        :class:`argparse.ArgumentParser` and use it to define extra
        arguments and options. These options will only affect the specified
        command.

        """

        def wrapper(func):
            self.add_command(func, *args, **kwargs)
            return func

        return wrapper

    def load_commands(self, obj):
        """
        Load commands defined on an arbitrary object.

        All functions decorated with the :func:`subparse.command` decorator
        attached the specified object will be loaded. The object may
        be a dictionary, an arbitrary python object, or a dotted path.

        The dotted path may be absolute, or relative to the current package
        by specifying a leading '.' (e.g. ``'.commands'``).

        """
        if isinstance(obj, str):
            obj = obj.replace(':', '.')
            package = caller_package().__name__ if obj.startswith('.') else None
            obj = import_module(obj, package)
        command.discover_and_call(obj, self.command)

    def load_commands_from_entry_point(self, specifier):
        """
        Load commands defined within a distribution entry point.

        Each entry will be a module that should be searched for functions
        decorated with the :func:`subparse.command` decorator. This
        operation is not recursive.

        """
        eps = importlib.metadata.entry_points()

        # getitem is deprecated in 3.10, so test for select and fallback
        # gracefully for 3.8/3.9
        entries = (
            eps.select(group=specifier) if hasattr(eps, 'select') else eps[specifier]
        )

        for ep in entries:
            module = ep.load()
            command.discover_and_call(module, self.command)

    def run(self, argv=None):
        """
        Run the command-line application.

        This will dispatch to the specified function or raise a
        ``SystemExit`` and output the appropriate usage information
        if there is an error parsing the arguments.

        The default ``argv`` is equivalent to ``sys.argv[1:]``.

        """
        if argv is None:  # pragma: no cover
            argv = sys.argv[1:]
        argv = [str(v) for v in argv]
        meta, args = parse_args(self, argv)
        context_factory = contextmanager(make_generator(self.context_factory))
        with context_factory(self, args, **meta.context_kwargs) as context:
            main = load_main(meta)
            return main(context, args) or 0


def parse_args(cli, argv):
    parser = cli._ArgumentParser(
        prog=cli.prog,
        usage=cli.usage,
        description=cli.description,
        formatter_class=argparse.RawTextHelpFormatter,
    )
    add_generic_options(parser, cli.generic_options)
    add_commands(parser, cli.commands, cli._namespace_key)
    try_argcomplete(parser)
    try:
        if cli.add_help_command:
            if argv and argv[0] == 'help':
                argv.pop(0)
                argv.append('--help')

        args = parser.parse_args(argv)
        meta = getattr(args, cli._namespace_key, None)
        if not meta:
            parser.print_help(file=sys.stderr)
            parser.exit(2)
        return meta, args

    except argparse.ArgumentError as e:
        parser.print_help(file=sys.stderr)
        parser.exit(2, f'{parser.prog}: error: {str(e)}\n')


def load_main(meta):
    main = meta.main
    if isinstance(main, str):
        mod = main
        func = 'main'
        if ':' in mod:
            mod, func = mod.split(':')
        mod = __import__(mod, None, None, ['__doc__'])
        main = getattr(mod, func)
    return main


def make_generator(fn):
    if inspect.isgeneratorfunction(fn):
        return fn

    def wrapper(*a, **kw):
        ctx = None
        if fn is not None:
            ctx = fn(*a, **kw)
        yield ctx

    return wrapper


def trim(docstring):
    """Trim function from PEP-257."""
    if not docstring:  # pragma: no cover
        return ''
    # Convert tabs to spaces (following the normal Python rules)
    # and split into a list of lines:
    lines = docstring.expandtabs().splitlines()
    # Determine minimum indentation (first line doesn't count):
    indent = sys.maxsize
    for line in lines[1:]:
        stripped = line.lstrip()
        if stripped:
            indent = min(indent, len(line) - len(stripped))
    # Remove indentation (first line is special):
    trimmed = [lines[0].strip()]
    if indent < sys.maxsize:
        for line in lines[1:]:
            trimmed.append(line[indent:].rstrip())
    # Strip off trailing and leading blank lines:
    while trimmed and not trimmed[-1]:
        trimmed.pop()
    while trimmed and not trimmed[0]:
        trimmed.pop(0)
    # Return a single string:
    return '\n'.join(trimmed)


def parse_docstring(docstring):
    """
    Parse a PEP-257 docstring.

    SHORT -> blank line -> LONG

    """
    short_desc = long_desc = ''
    if docstring:
        docstring = trim(docstring)
        lines = docstring.split('\n\n', 1)
        short_desc = lines[0].strip().replace('\n', ' ')

        if len(lines) > 1:
            long_desc = lines[1].strip()
    return short_desc, long_desc


def try_argcomplete(parser):  # pragma: no cover
    try:
        import argcomplete
    except ImportError:
        pass
    else:
        argcomplete.autocomplete(parser)


def add_generic_options(parser, fns):
    for func in fns:
        func(parser)


def add_commands(parser, commands, namespace_key):
    subparsers = parser.add_subparsers(title='commands', metavar='<command>')
    for meta in sorted(commands.values(), key=lambda m: m.name):
        subparser = subparsers.add_parser(
            meta.name,
            help=meta.help,
            description=meta.description,
            formatter_class=argparse.RawTextHelpFormatter,
        )
        meta.factory(subparser)
        subparser.set_defaults(**{namespace_key: meta})


# stolen from pyramid.path
def caller_module(level=2):
    module_globals = sys._getframe(level).f_globals
    module_name = module_globals.get('__name__') or '__main__'
    module = sys.modules[module_name]
    return module


# stolen from pyramid.path
def package_for_module(module):  # pragma: no cover
    f = getattr(module, '__file__', '')
    if ('__init__.py' in f) or ('__init__$py' in f):  # empty at >>>
        # Module is a package
        return module
    # Go up one level to get package
    package_name = module.__name__.rsplit('.', 1)[0]
    return sys.modules[package_name]


# stolen from pyramid.path
def caller_package(level=2):
    # caller_module in arglist for tests
    module = caller_module(level + 1)
    return package_for_module(module)
